Mapping guide doc (#131)
FAIRmat-NFDI/nomad-parser-plugins-simulation: docs/mapping-annotations-for-AreaC.md

# Practical Guide to Mapping Annotations for Parser dev

This doc explains how to populate the NOMAD simulations datamodel using mapping annotations. It shows what they are, how to write them, and how to handle nested and repeating sections. 

---

## What a mapping annotation is

> - A small note on a schema field that says: *for source `<KEY>`, read value at `<PATH>` and put it here.*

Reviewer: ```suggestion
- A small note on a schema field that says: *for source `<KEY>`, read value at `<PATH>` and place it in the archive under the path leading to the annotated schema field.*
```

Author: this is what i initially wanted to avoid actually, as this internal document is trying to have a casual and simple approach

- Stored on any section/quantity: `m_annotations['mapping'][<KEY>] = MapperAnnotation(mapper=...)`.

> - `<KEY>` names the source (e.g., `OUT_KEY`, `HDF5_KEY`, `INFO_KEY`, `BAND_KEY`).

Reviewer: I would add "why we there may be multiple keys". If there was only one, that field would not be necessary.

Author: this part is focusing on what, and not why. we should avoid giving out too many details at once

Reviewer: sure, then maybe add the "why" above

> - `<KEY>` names the source (e.g., `OUT_KEY`, `HDF5_KEY`, `INFO_KEY`, `BAND_KEY`).

Reviewer: Minor nitpick: it doesn't "name" the source per se (as in it does not define the source file).
It just defines a particular source format / mapping.

Author: will revise it :+1:

> - Paths use JMESPath style. Relative paths start with `.` (from the parent’s mapped node). `@` anchors at the source root.

Reviewer: ```suggestion
- Paths use [JMESPath](https://jmespath.org/) style. Relative paths start with `.` (from the parent’s mapped node). `@` anchors at the source root.
```

- The machinery (`MetainfoParser`) walks the schema, follows these notes, instantiates sections/subsections, and fills values.

## What happens under the hood

1) A source parser (`XMLParser`, `HDF5Parser`, `TextParser`, etc.) reads a file into a Python dict.
2) Set `annotation_key` on a `MetainfoParser` to the `<KEY>` that matches the annotations.

> 3) Call `convert(source_parser, target_parser)`. One key → one `convert`. Multiple keys/sources → run `convert` once per key after switching `annotation_key` each time.

Reviewer: > One key → one `convert`.

Lol, sounds oddly religious.

4) The target archive is populated; repeating subsections are created automatically from lists.

## Advantages of using mapping

- Declarative and readable: data movement is described alongside the schema, not scattered in code.
- Reusable across sources: same schema, different annotation keys; swap sources by switching `annotation_key`.

> - Automatic instantiation: nested and repeating subsections are created based on list structure; no manual wiring.

Reviewer: Could you clarify what you mean by "nested" here? It's a very loaded term given the ES component.

I'm asking, as there is a special edge case, but I don't get the feeling that is what you are aiming at here. I simply read it as "a subsection" here, i.e.

```suggestion
- Automatic instantiation of repeating sections and subsections; no manual wiring.
```

> - Composable transforms: reshape, filter, and unit application via small helper functions in the parser.

Reviewer: Agreed. Since you mentioned declarative style above, and these applications are (conceptually) immutable, you could even tentatively call it "functional" here. Jives really well with composition too.

- Easier maintenance: when input formats change, update paths in one place (annotations) instead of parser code.
- Parallel branches: ingest multiple files/sources into the same archive by running `convert` per key.

## Path cheat sheet

> - Relative from parent: `.field`, `.field[0]`, `.list[*].value`

Reviewer: but no index formula, unfortunately

- Absolute from root: `field.subfield` (no leading dot)
- Root anchor: `@`
- Filter: `.items[?"@name"=="version"] | [0].__value`

> - Attributes (XML/HDF5): keys often prefixed with `@`; values with `__value`

Reviewer: This needs an example to distinguish from the root anchor.

> - Attributes (XML/HDF5): keys often prefixed with `@`; values with `__value`

Reviewer 2: Does "often" here mean that this is a convention, but it would theoretically also be possible, to, e.g., prefix a value with `@`?

## When to annotate `m_def` vs. the attribute

- Attribute-level (on the quantity or subsection attribute):

>   - Use for a specific, local mapping. This is the default for quantities. Example: `Program.version` with `.program_version`.

Reviewer: Local mapping means that it isn't inherited from, correct? I.e. "standalone use"

>   - Use for a specific, local mapping. This is the default for quantities. Example: `Program.version` with `.program_version`.

Reviewer: Maybe clarify that `Program.version` is the schema, while `.program_version` is the extracted path.

  - It overrides `m_def` if both exist.
- `m_def` (on the section definition):
  - Use for a default that should work anywhere that section appears. Example: `Simulation.m_def` with `@` to anchor the whole tree.
  - Use to give a subsection its own base path so children can resolve relative paths. Example: `DFT.m_def` with `.dft`, so all quantities inside DFT can start from `.dft` even if the parent doesn’t have a special annotation.
  - Use for abstract/base sections when you need a mapping that applies to all their concrete inheritors.
- Lookup order: attribute annotation → `m_def` annotation → inheriting section definitions (for abstract subsections). If you want a local override, put it on the attribute; if you want a reusable default or a base path, put it on `m_def`.

## Core patterns 

### 1) Root anchor
Anchor the target root section to the **source tree root** (`@` = top of source dict/XML/HDF5).
```python
add_mapping_annotation(general.Simulation.m_def, OUT_KEY, '@')
```

> ### 2) Nested quantities

Reviewer: ```suggestion
### 2) Quantities
```

Quantities can never be at the root.

Relative to the parent section’s mapped node.
```python
add_mapping_annotation(general.Program.name, OUT_KEY, '.program')
add_mapping_annotation(general.Program.version, OUT_KEY, '.program_version')
```

> ### 3) Single nested subsection

Reviewer: ```suggestion
### 3) Single subsection
```

Tautology: the prefix "sub" already clarifies that is "nested" as you refer to it. Moreover, "nested" obtains a very different meaning when talking about serialization and querying. Would avoid it to prevent confusion.

Annotate the subsection’s section definition.
```python
add_mapping_annotation(model_method.DFT.m_def, OUT_KEY, '.dft')
```

> ### 4) Repeating subsections (lists)

Reviewer: ```suggestion
### 4) Repeating subsections & Array-like Quantities
```

Point to a list; one subsection is created per element. For repeating quantities or array-like quantities, map to a list/array; it is set directly.
```python
add_mapping_annotation(model_system.AtomsState.m_def, OUT_KEY, '.atoms')
add_mapping_annotation(model_system.AtomsState.chemical_symbol, OUT_KEY, '.label')
add_mapping_annotation(model_system.AtomsState.atomic_number, OUT_KEY, '.number')
add_mapping_annotation(model_system.ModelSystem.positions, OUT_KEY, '.positions')  # shape [n_atoms, 3]
```

### 5) Multiple sources into the same schema
Use different keys; run `convert` once per key.
```python
# Octopus uses OUT_KEY + INFO_KEY + EIGENVALUES_KEY on the same schema
# Exciting uses INFO_KEY, INPUT_XML_KEY, EIGVAL_KEY, BANDSTRUCTURE_XML_KEY, DOS_XML_KEY
# Wannier90 uses WOUT_KEY, WIN_KEY, BAND_KEY, WHR_KEY, DOS_KEY
```

> Set `annotation_key = <KEY>` before each `convert`.

Reviewer: Can you add an example, pls?
From memory, I'd say that `convert` is called at the parser-level instead of the annotations, but might be fuzzy on that one.

### 6) Transformations (reshape, filter, apply units)

> Use a tuple `(function_name, [paths], {kwargs})` and implement the function on the parser.

Reviewer: ```suggestion
Use a tuple `(function_name, [paths], {kwargs})` and implement the function at the parser level.
```

Maybe you should introduce the 2 levels.

```python
add_mapping_annotation(
    outputs.TotalForce.value,
    SECONDARY_KEY,
    ('reshape_forces', ['.raw_forces'], {'unit': 'eV / angstrom'})
)
```

### 7) Units or subfield search
You can pass `unit='eV'` or `search='.path.to.value'` in the annotation (see Wannier90 eigenvalue/DOS paths).

## Starter template

> File: `packages/nomad_simulation_parsers/schema_packages/template.py`

Reviewer: What is the purpose of this line? Regardless, given the size of this template, I would move it out to exactly that file path. You should of course reference it from these docs.

add_mapping_annotation(general.Program.version, PRIMARY_KEY, '.program.version')

# Single nested subsection
add_mapping_annotation(model_system.ModelSystem.m_def, PRIMARY_KEY, '.structure')
add_mapping_annotation(model_system.ModelSystem.lattice_vectors, PRIMARY_KEY, '.lattice')
add_mapping_annotation(model_system.ModelSystem.periodic_boundary_conditions, PRIMARY_KEY, '.pbc')

# Repeating subsections
add_mapping_annotation(model_system.AtomsState.m_def, PRIMARY_KEY, '.atoms')
add_mapping_annotation(model_system.AtomsState.chemical_symbol, PRIMARY_KEY, '.symbol')
add_mapping_annotation(model_system.AtomsState.atomic_number, PRIMARY_KEY, '.Z')
add_mapping_annotation(model_system.ModelSystem.positions, PRIMARY_KEY, '.positions')

# Repeating outputs
add_mapping_annotation(outputs.Outputs.total_energies, PRIMARY_KEY, '.energies')
add_mapping_annotation(outputs.TotalEnergy.value, PRIMARY_KEY, '.value')
add_mapping_annotation(outputs.TotalEnergy.name, PRIMARY_KEY, '.label')

# Transformation example on secondary source
add_mapping_annotation(
    outputs.TotalForce.value,
    SECONDARY_KEY,
    ('reshape_forces', ['.raw_forces'], {'unit': 'eV / angstrom'}),
)
```

> File: `packages/nomad_simulation_parsers/parsers/template/parser.py`

Reviewer: Same comment as on the schema-side

Author: it is a brainstorm that we might have example template files stored in the parsers repo to guide people

Reviewer: Sure. The template is good. I would just move it its own file already.

        self._source_primary.filepath = mainfile
        self._source_secondary.filepath = mainfile.replace('.out', '.aux')  # adjust

        self.archive = Simulation(program=Program(name='TEMPLATE'))
        self._target.data_object = self.archive

        self._target.annotation_key = template.PRIMARY_KEY
        self._source_primary.convert(self._target)

        self._target.annotation_key = template.SECONDARY_KEY
        self._source_secondary.convert(self._target)

        return self.archive
```

## How to add a new mapping (step-by-step)

1) Pick the target field in `nomad-simulations` (quantity or subsection).
2) Choose a source key (`OUT_KEY`, `HDF5_KEY`, etc.).
3) Write the JMESPath that matches your source dict. Use `.` to start from the parent node.
4) Add the annotation with `add_mapping_annotation(target, KEY, path_or_tuple)`.
5) Repeats: if the subsection is `repeats=True`, just point to a list path; one instance per list element.
6) If combining multiple sources, run `convert` for each key (set `annotation_key` before calling).
7) Test on a small file, inspect the archive (e.g., print `archive.data`), and adjust paths.

## Pitfalls to avoid

> - Forgetting the leading `.` on relative paths (they become absolute and miss).

Reviewer: So, I don't think this gets covered anywhere, but you raise a valid point:
`.` for relative, `@` for from the root, but what if there is no special character at the front? Where does that path start from?

> Forgetting the leading `.`

I would put the answer higher up. Then this pitfall point can be dropped.

> - Abstract subsections: put annotations on concrete inheriting sections or rely on inheriting annotations.

Reviewer: > Abstract subsections

"Abstract subsections" are those inheriting from `ABC`. Upon re-reading, I guess you mean the imparitive verb mode. If so:

```suggestion
- Avoid annotating every subsection in the chain: put annotations on concrete inheriting sections or rely on inheriting annotations.
```

Still feels like this could become its own section.

> - Reusing a schema with different sources: clear old mappings first with `remove_mapping_annotations`.

Reviewer: When exactly should `remove_mapping_annotations` be called?
